File: src/main/scala/kr/ac/kaist/jiset/js/ast/PropertyDefinition.scala
package kr.ac.kaist.jiset.js.ast

import kr.ac.kaist.jiset.ir._
import kr.ac.kaist.jiset.util.Span
import kr.ac.kaist.jiset.util.Useful._
import io.circe._, io.circe.syntax._

trait PropertyDefinition extends AST { val kind: String = "PropertyDefinition" }

object PropertyDefinition {
  def apply(data: Json): PropertyDefinition = AST(data) match {
    case Some(compressed) => PropertyDefinition(compressed)
    case None => error("invalid AST data: $data")
  }
  def apply(data: AST.Compressed): PropertyDefinition = {
    val AST.NormalCompressed(idx, subs, params, span) = data
    idx match {
      case 0 =>
        val x0 = subs(0).map(IdentifierReference(_)).get
        PropertyDefinition0(x0, params, span)
      case 1 =>
        val x0 = subs(0).map(CoverInitializedName(_)).get
        PropertyDefinition1(x0, params, span)
      case 2 =>
        val x0 = subs(0).map(PropertyName(_)).get
        val x1 = subs(1).map(AssignmentExpression(_)).get
        PropertyDefinition2(x0, x1, params, span)
      case 3 =>
        val x0 = subs(0).map(MethodDefinition(_)).get
        PropertyDefinition3(x0, params, span)
      case 4 =>
        val x0 = subs(0).map(AssignmentExpression(_)).get
        PropertyDefinition4(x0, params, span)
    }
  }
}

case class PropertyDefinition0(x0: IdentifierReference, parserParams: List[Boolean], span: Span) extends PropertyDefinition {
  x0.parent = Some(this)
  def idx: Int = 0
  def k: Int = d(x0, 0)
  def fullList: List[(String, PureValue)] = l("IdentifierReference", x0, Nil).reverse
  def maxK: Int = 0
  override def toString: String = {
    s"$x0"
  }
}

case class PropertyDefinition1(x0: CoverInitializedName, parserParams: List[Boolean], span: Span) extends PropertyDefinition {
  x0.parent = Some(this)
  def idx: Int = 1
  def k: Int = d(x0, 0)
  def fullList: List[(String, PureValue)] = l("CoverInitializedName", x0, Nil).reverse
  def maxK: Int = 0
  override def toString: String = {
    s"$x0"
  }
}

case class PropertyDefinition2(x0: PropertyName, x2: AssignmentExpression, parserParams: List[Boolean], span: Span) extends PropertyDefinition {
  x0.parent = Some(this)
  x2.parent = Some(this)
  def idx: Int = 2
  def k: Int = d(x2, d(x0, 0))
  def fullList: List[(String, PureValue)] = l("AssignmentExpression", x2, l("PropertyName", x0, Nil)).reverse
  def maxK: Int = 0
  override def toString: String = {
    s"$x0 : $x2"
  }
}

case class PropertyDefinition3(x0: MethodDefinition, parserParams: List[Boolean], span: Span) extends PropertyDefinition {
  x0.parent = Some(this)
  def idx: Int = 3
  def k: Int = d(x0, 0)
  def fullList: List[(String, PureValue)] = l("MethodDefinition", x0, Nil).reverse
  def maxK: Int = 0
  override def toString: String = {
    s"$x0"
  }
}

case class PropertyDefinition4(x1: AssignmentExpression, parserParams: List[Boolean], span: Span) extends PropertyDefinition {
  x1.parent = Some(this)
  def idx: Int = 4
  def k: Int = d(x1, 0)
  def fullList: List[(String, PureValue)] = l("AssignmentExpression", x1, Nil).reverse
  def maxK: Int = 0
  override def toString: String = {
    s"... $x1"
  }
}

File: lib/esparse/src/es2021/PropertyDefinition.js
const Node = require('../Node');

// PropertyDefinition[Yield, Await] :
//    IdentifierReference[?Yield, ?Await]
//    CoverInitializedName[?Yield, ?Await]
//    PropertyName[?Yield, ?Await] : AssignmentExpression[+In, ?Yield, ?Await]
//    MethodDefinition[?Yield, ?Await]
//    ... AssignmentExpression[+In, ?Yield, ?Await]
let PropertyDefinition = (Yield, Await) => (given) => {
  let IdentifierReference = require('./IdentifierReference');
  let CoverInitializedName = require('./CoverInitializedName');
  let PropertyName = require('./PropertyName');
  let MethodDefinition = require('./MethodDefinition');

  let AssignmentExpression = require('./AssignmentExpression');
  let params = [Yield, Await];
  let { method, shorthand, key, value } = given;
  
  switch (given.kind) {
    case 'get': 
    case 'set': {
      let method = MethodDefinition(Yield, Await)(given);
      return new Node('PropertyDefinition', given, 3, [method], params);
    }
    case 'init': {
      if (shorthand && value.type === 'AssignmentPattern') {
        let cov = CoverInitializedName(Yield, Await)(value);
        return new Node('PropertyDefinition', given, 1, [cov], params);
      } else if (shorthand) {
        let ref = IdentifierReference(Yield, Await)(key);
        return new Node('PropertyDefinition', given, 0, [ref], params);
      } else if (!method) {
        let name = PropertyName(Yield, Await)(given);
        let expr = AssignmentExpression(true, Yield, Await)(value);
        return new Node('PropertyDefinition', given, 2, [name, expr], params);
      } else if (method) {
        let method = MethodDefinition(Yield, Await)(given);
        return new Node('PropertyDefinition', given, 3, [method], params);
      }
      return Node.TODO(`init @ PropertyDefinition`);
    }
    default:
      Node.TODO(`${given.kind} @ PropertyDefinition`);
  }
}

module.exports = PropertyDefinition;
